src/ape/managers/chain.py

```diff
@@ -377,89 +383,113 @@ def _get_block(self, block_id: BlockID) -> BlockAPI:
         return self.provider.get_block(block_id)
 
 
-class AccountHistory(BaseManager):
+class AccountHistory(BaseInterfaceModel):
     """
     A container mapping account addresses to the transaction from the active session.
     """
 
-    _address_list_map: Dict[AddressType, List[ReceiptAPI]] = {}
-
-    # Facilitates faster look-ups.
-    _hash_to_receipt_map: Dict[str, ReceiptAPI] = {}
+    address: AddressType
+    sessional: List[ReceiptAPI] = []
 
-    def __getitem__(self, address: Union[BaseAddress, AddressType, str]) -> List[ReceiptAPI]:
+    @property
+    def outgoing(self) -> Iterator[ReceiptAPI]:
         """
-        Get the list of transactions from the active session for the given address.
-
-        Args:
-            address (``AddressType``): The sender of the desired transactions.
-
-        Returns:
-            List[:class:`~ape.api.transactions.TransactionAPI`]: The list of transactions. If there
-            are no recorded transactions, returns an empty list.
+        All outgoing transactions, from earliest to latest.
         """
-
-        address_key: AddressType = self.conversion_manager.convert(address, AddressType)
         explorer = self.provider.network.explorer
         explorer_receipts = (
-            [r for r in explorer.get_account_transactions(address_key)] if explorer else []
+            [r for r in explorer.get_account_transactions(self.address)] if explorer else []
         )
         for receipt in explorer_receipts:
+            if receipt.sender != self.address:
+                # Likely ``self.address`` is a contract.
+                # Cache the receipts by their sender instead and skip them here.
+                self.chain_manager.history.append(receipt)
+                continue
 
-            # NOTE: Cache the receipt by its sender and not by the address sent to the explorer API
-            #  because explorers return various receipts when given contract addresses.
-            sender: AddressType = receipt.sender
-            if receipt.txn_hash not in [r.txn_hash for r in self._address_list_map.get(sender, [])]:
-                self.append(receipt)
+            yield receipt
 
-        return self._address_list_map.get(address_key, [])
+    def __iter__(self) -> Iterator[ReceiptAPI]:  # type: ignore[override]
+        yield from self.outgoing
 
-    def __iter__(self) -> Iterator[AddressType]:
-        """
-        Iterate through the accounts listed in the history map.
+    def __len__(self) -> int:
+        return len(list(self.outgoing))
```

Comment:
This can just be `self.provider.get_nonce(self.address)`

Reply:
or `nonce + 1`, not sure if it's the next nonce or the last one

Reply:
I love this!
But one thing to note is that it could be ecosystem dependent regarding whether it starts at 0 or 1, lol. I think Starknet was different at one point.

Reply:
But for Ethereum it would be `account.nonce` I think because it starts as `0` and you haven't sent any txns yet.
I am not sure about contracts though.

Reply:
Worth following up on

Wonder if starknet started at 1 because accounts needed to be deployed first

Reply:
> I am not sure about contracts though.

contracts have no ability to create outgoing transactions (defined as transactions where they are original source)
https://etherscan.io/txs?a=0xfeb4acf3df3cdea7399794d0869ef76a6efaff52&f=2

Reply:
So for contracts, you won't be able to search their history using `.outgoing`, only by using `.incoming` or `.internal` (which are much heavier queries to make)